README.md
# ABOUT

cockpit.js lets you add an retro-style 2D cockpit overlay to your 3D game and
lets you move it around, vibrate and shake it to simulate the player's head
movement and/or physical impact on the ship/car/plane/vehicle.

The idea is strikingly simple and has worked for years in simulators (and is
still being used today in games). I figured that the implementation would be
strikingly simple as well and gave it a shot on a sunday night. But, I like it,
and plan to add more features to it without increasing it's low complexity.

# Usage

Grab the cockpit.js file and include it in your page.

To add a cockpit to your game, just create a new Cockpit instance and pass the
path to the cockpit overlay image:

`var cockpit = new Cockpit('path/to/my/overlay.png');`

The most basic thing to do now is to move your cockpit around:

`cockpit.move(horizontalValue, verticalValue);`

The two values you have to pass must be numbers ranging from 0 to 1, where 1 is  the maximum movement.

Next, make your cockpit vibrate for a set duration:

`cockpit.vibrate(duration);`

Duration is passed in ms and optional, defaulting to 600ms.

If you don't now upfront how long your cockpit is going to vibrate, use the following two methods:

`cockpit.beginVibrate();` and `cockpit.endVibrate();`

If things go hairy, e.g. your vehicle is being hit, you can shake your cockpit:

`cockpit.shake(intensity, duration);`

Where intensity is a number from 0 to 1. Duration is passed in ms. Both arguments are optional, intensity defaulting to 0.5 and duration to 300.

As with vibrating, you can shake the cockpit for an unkown amout of time using

`cockpit.beginShake(intensity);` and `cockpit.endShake();`

The intensity param is optional and works as described above.

Instead of the begin/end methods, you can also use the following method:

`cockpit.setState(state);`

where state is one of 'normal', 'vibrate' or 'shake'. You can't pass any options for now (it will use default values), but you have a convenient method to toggle between the different states.

You can also add some HUD-like text items to your cockpit, that will move around with the cockpit:

`cockpit.addText(id, textToDisplay);`

The id is applied to the text item's DOMNode, so you can use it to position and style it. For positioning, I recommend to use absolute positioning with percentages so that the text will stay in place independent of window size (well, almost, soo TODO section).

To update your text item, just call

`cockpit.updateText(id, textToDisplay);`

with the id you previously passed and the new text.


# Demos

* [Using cockpit.js over a three.js example.](http://jensarps.github.com/cockpit.js/demo/three) When you get too close to the earth's atmosphere, the ship will start to vibrate and lose speed. Observe the HUD to see your current thrust, speed and g-force.
* [A cockpit over a white background.](http://jensarps.github.com/cockpit.js/demo/plain) Yeah. Just for a quick test. Doubleclick to shake, right click to vibrate.


# TODO

* Use transform3d() instead of left/right and compare performance
* ~~Add beginShake() and endShake() methods to do ongoing shaking~~
* ~~Add methods to conveniently add HUD like text to the cockpit~~
* Add methods to allow grouping of text items
* Handle positioning of text inside of the code to make it precisely show up at the desired spot
* Add preventMove param to vibrate and shake to enable more precise movements
* Change shake and vibrate signature to kwArgs to not screw up existing code
when new params are introduced
* Add z-axis movement, might look nice for intense de-/acceleration
* Add UMD wrapper


CHANGELOG
0.1
* Initial release

0.2
* Added addText and updateText methods

0.3
* Added beginShake and endShake methods
* Introduced states, added setState method
* Performance tweaks


demo/three/SpaceshipControls.js
/**
 * @author Jens Arps
 *
 * Based on FlyControls by James Baicoianu / http://www.baicoianu.com/
 */

THREE.SpacheshipControls = function ( object, domElement ) {

	this.object = object;

	this.domElement = ( domElement !== undefined ) ? domElement : document;
	if ( domElement ) this.domElement.setAttribute( 'tabindex', -1 );

  // initial values
	this.movementSpeed = 0;
  this.velocity = 0;

  // influence
  this.breakingForce = 0;

  // ship stats
  this.maxSpeed = 1000;
  this.inertia = 100;
  this.rollSpeed = 0.005;

	this.autoForward = false;

	// disable default target object behavior

	this.object.useQuaternion = true;

	// internals

	this.tmpQuaternion = new THREE.Quaternion();

	this.mouseStatus = 0;

	this.moveState = {
    up: 0,
    down: 0,
    left: 0,
    right: 0,
    forward: 1,
    back: 0,
    pitchUp: 0,
    pitchDown: 0,
    yawLeft: 0,
    yawRight: 0,
    rollLeft: 0,
    rollRight: 0
  };
	this.moveVector = new THREE.Vector3( 0, 0, 0 );
	this.rotationVector = new THREE.Vector3( 0, 0, 0 );

	this.keydown = function( event ) {

		if ( event.altKey ) {

			return;

		}

		switch( event.keyCode ) {

      case 87: /*W*/ this.velocity = Math.min(1, this.velocity + 0.015); break;
      case 83: /*S*/ this.velocity = Math.max(0, this.velocity - 0.015); break;

			// case 65: /*A*/ this.moveState.left = 1; break;
			// case 68: /*D*/ this.moveState.right = 1; break;

			// case 82: /*R*/ this.moveState.up = 1; break;
			// case 70: /*F*/ this.moveState.down = 1; break;

			// case 37: /*left*/ this.moveState.yawLeft = 1; break;
			// case 39: /*right*/ this.moveState.yawRight = 1; break;

			case 81: /*Q*/ this.moveState.rollLeft = 1; break;
			case 69: /*E*/ this.moveState.rollRight = 1; break;

		}

		this.updateMovementVector();
		this.updateRotationVector();

	};

  this.keyup = function( event ) {

  		switch( event.keyCode ) {

  			// case 87: /*W*/ this.moveState.forward = 0; break;
  			// case 83: /*S*/ this.moveState.back = 0; break;

  			// case 65: /*A*/ this.moveState.left = 0; break;
  			// case 68: /*D*/ this.moveState.right = 0; break;

  			// case 82: /*R*/ this.moveState.up = 0; break;
  			// case 70: /*F*/ this.moveState.down = 0; break;

  			// case 38: /*up*/ this.moveState.pitchUp = 0; break;
  			// case 40: /*down*/ this.moveState.pitchDown = 0; break;

  			// case 37: /*left*/ this.moveState.yawLeft = 0; break;
  			// case 39: /*right*/ this.moveState.yawRight = 0; break;

  			case 81: /*Q*/ this.moveState.rollLeft = 0; break;
  			case 69: /*E*/ this.moveState.rollRight = 0; break;

  		}

  		this.updateMovementVector();
  		this.updateRotationVector();

  	};

	this.mousemove = function( event ) {

			var container = this.containerDimensions;
			var halfWidth  = container.size[ 0 ] / 2;
			var halfHeight = container.size[ 1 ] / 2;

			this.moveState.yawLeft   = - ( ( event.pageX - container.offset[ 0 ] ) - halfWidth  ) / halfWidth;
			this.moveState.pitchDown = - ( ( event.pageY - container.offset[ 1 ] ) - halfHeight ) / halfHeight;

			this.updateRotationVector();

	};

	this.update = function( delta ) {

    var factor = Math.max(0, this.velocity - this.breakingForce);
    var diff = factor - this.movementSpeed;
    this.movementSpeed += (diff / this.inertia);

    var speed = this.movementSpeed * this.maxSpeed;

		var moveMult = delta * speed;
		var rotMult = delta * this.rollSpeed;

		this.object.translateX( this.moveVector.x * moveMult );
		this.object.translateY( this.moveVector.y * moveMult );
		this.object.translateZ( this.moveVector.z * moveMult );

		this.tmpQuaternion.set( this.rotationVector.x * rotMult, this.rotationVector.y * rotMult, this.rotationVector.z * rotMult, 1 ).normalize();
		this.object.quaternion.multiplySelf( this.tmpQuaternion );

		this.object.matrix.setPosition( this.object.position );
		this.object.matrix.setRotationFromQuaternion( this.object.quaternion );
		this.object.matrixWorldNeedsUpdate = true;


	};

	this.updateMovementVector = function() {

		var forward = ( this.moveState.forward || ( this.autoForward && !this.moveState.back ) ) ? 1 : 0;

		this.moveVector.x = ( -this.moveState.left    + this.moveState.right );
		this.moveVector.y = ( -this.moveState.down    + this.moveState.up );
		this.moveVector.z = ( -forward + this.moveState.back );

		//console.log( 'move:', [ this.moveVector.x, this.moveVector.y, this.moveVector.z ] );

	};

	this.updateRotationVector = function() {

		this.rotationVector.x = ( -this.moveState.pitchDown + this.moveState.pitchUp );
		this.rotationVector.y = ( -this.moveState.yawRight  + this.moveState.yawLeft );
		this.rotationVector.z = ( -this.moveState.rollRight + this.moveState.rollLeft );

		//console.log( 'rotate:', [ this.rotationVector.x, this.rotationVector.y, this.rotationVector.z ] );

	};

	this.onContainerDimensionsChanged = function() {

    var dimensions;

		if ( this.domElement != document ) {
			dimensions = {
				size	: [ this.domElement.offsetWidth, this.domElement.offsetHeight ],
				offset	: [ this.domElement.offsetLeft,  this.domElement.offsetTop ]
			};
		} else {
			dimensions = {
				size	: [ window.innerWidth, window.innerHeight ],
				offset	: [ 0, 0 ]
			};
		}
    this.containerDimensions = dimensions;
	};

	function bind( scope, fn ) {
		return function () {
			fn.apply( scope, arguments );
		};
	};

	this.domElement.addEventListener( 'mousemove', bind( this, this.mousemove ), false );
	this.domElement.addEventListener( 'keydown', bind( this, this.keydown ), false );
  this.domElement.addEventListener( 'keyup', bind( this, this.keyup ), false );

	this.updateMovementVector();
	this.updateRotationVector();
  this.onContainerDimensionsChanged();

};


cockpit.js

;(function(global){

  var Cockpit = function(url){
    this.imageUrl = url;

    this.setup();
  };

  Cockpit.prototype = {

    imageUrl: null,

    cockpitNode: null,

    wrapperNode: null,

    textNodes: null,

    currentY: 0,

    currentX: 0,

    currentZ: 0,

    state: 'normal',

    setup: function() {
      this.textNodes = {};
      this.setupNodes();
      this.textProperty = typeof this.wrapperNode.innerText !== 'undefined' ? 'innerText' : 'textContent';
      this.translateProperty = 'webkitTransform' in this.wrapperNode.style ? 'webkitTransform' : 'mozTransform'
    },

    setupNodes: function() {
      var wrapperNode = this.wrapperNode = document.createElement('div');
      var wrapperNodeStyle = wrapperNode.style;
      wrapperNodeStyle.width = '100%';
      wrapperNodeStyle.height = '100%';
      wrapperNodeStyle.position = 'absolute';
      wrapperNodeStyle.left = '0';
      wrapperNodeStyle.top = '0';
      wrapperNodeStyle.overflow = 'hidden';
      wrapperNodeStyle.zIndex = '100000';
      wrapperNodeStyle.pointerEvents = 'none';
      document.body.appendChild(wrapperNode);

      var cockpitNode = this.cockpitNode = document.createElement('div');
      var cockpitNodeStyle = cockpitNode.style;
      cockpitNodeStyle.background = 'url(' + this.imageUrl + ') center';
      cockpitNodeStyle.backgroundSize = 'cover';
      cockpitNodeStyle.width = '130%';
      cockpitNodeStyle.height = '130%';
      cockpitNodeStyle.position = 'absolute';
      cockpitNodeStyle.left = '-15%';
      cockpitNodeStyle.top = '-15%';
      wrapperNode.appendChild(cockpitNode);
    },

    /**
     * Moves the cockpit around.
     *
     * Pass two values, one for vertical movement and one for horizontal
     * movement. Range for these values is from -1 (max negative movement) to
     * 1 (max positive movement). 0 means no movement.
     *
     * @param {number} x The horizontal movement
     * @param {number} y The vertical movement
     * @param {number} z The depth movement
     */
    move: function(x, y, z){
      var isDirty = false;

      if (x != null && x !== this.currentX){
        isDirty = true;
        this.currentX = x;
      }

      if (y != null && y !== this.currentY){
        isDirty = true;
        this.currentY = y;
      }

      if (z != null && z !== this.currentZ){
        isDirty = true;
        this.currentZ = z;
      }

      if(isDirty){
        var cockpitNodeStyle = this.cockpitNode.style;
        var left = ( this.currentX * 7.5).toFixed(3) + '%';
        var top = ( this.currentY * 7.5).toFixed(3) + '%';
        var scale = this.currentZ * 0.5 + 1;
        cockpitNodeStyle[this.translateProperty] = 'translate3d(' + left + ',' + top + ',0) scale(' + scale + ')';
      }
    },

    /**
     * Creates a random position around current position
     *
     * @param {number} current The current position value
     * @param {number} intensity Thr random intensity, from 0 to 1, where 0
     *    equals no randomization and 1 equals full randomization
     */
    getRandomPosition: function(current, intensity){
      return Math.max(Math.min((( Math.random() - 0.5 ) * intensity * 0.5) + current, 1.5), -1.5);
    },

    /**
     * Shakes the cockpit.
     *
     * You can pass an optional intensity value from 0 to 1 (defaults to 0.5)
     * and and optional duration value (defaults to 300).
     *
     * @param {number} [intensity] The shake intensity from 0 to 1
     * @param {number} [duration] The shake duration in ms
     */
    shake: function(intensity, duration){
      if(this.state === 'shake'){
        return;
      }
      intensity = intensity || 0.5;
      duration = duration || 300;

      setTimeout(this.endShake.bind(this), duration);
      this.beginShake(intensity);
    },

    /**
     * Starts to shake the cockpit
     *
     * The cockpit will shake until endShake() is called.
     *
     * @param {number} [intensity] The shake intensity from 0 to 1
     */
    beginShake: function(intensity){
      if(this.state === 'shake'){
        return;
      }
      this.state = 'shake';
      intensity = intensity || 0.5;
      var interval;

      interval = setInterval(function(){
        if (this.state !== 'shake') {
          clearInterval(interval);
          return;
        }
        this.move(this.getRandomPosition(this.currentX, intensity), this.getRandomPosition(this.currentY, intensity));
      }.bind(this), 50);
    },

    /**
     * Ends the shaking started with beginShake()
     *
     */
    endShake: function(){
      this.state = 'normal';
    },

    /**
     * Vibrates the cockpit for a set duration
     *
     * You can pass an optional duration value.
     *
     * @param {number} [duration] The vibration duration
     */
    vibrate: function(duration){
      if(this.state === 'vibrate'){
        return;
      }
      duration = duration || 600;

      setTimeout(this.endVibrate.bind(this), duration);
      this.beginVibrate();
    },

    /**
     * Starts to vibrate the cockpit
     *
     * The cockpit will vibrate until endVibrate() is called
     */
    beginVibrate: function(){
      if(this.state === 'vibrate'){
        return;
      }
      this.state = 'vibrate'
      var direction = 1;
      var interval;

      interval = setInterval(function(){
        if (this.state !== 'vibrate') {
          clearInterval(interval);
          return;
        }
        direction *= -1;
        var modificator = 0.01 * direction;
        this.move(this.currentX + modificator, this.currentY + modificator);
      }.bind(this), 30);
    },

    /**
     * Ends the vibration started by beginVibrate()
     *
     */
    endVibrate: function(){
      this.state = 'normal';
    },

    /**
     * Sets the state of the cockpit
     *
     * Will also end previous states if differs from current.
     *
     * This can be used instead of the begin/end methods.
     *
     * @param {string} state The new state to set, can be one of 'vibrate',
     *    'shake' or 'normal'
     */
    setState: function(state){
      if(this.state === state){
        return;
      }
      switch(state){
        case 'vibrate':
          this.beginVibrate();
          break;
        case 'shake':
          this.beginShake();
          break;
        default:
          this.state = 'normal';
          break;
      }
    },

    /**
     * Adds a text item to the cockpit overlay
     *
     * @param {string} id An identifier for this text item
     * @param {sring} initialValue An initial value to be displayed
     */
    addText: function(id, initialValue){
      var node = document.createElement('span');
      node[this.textProperty] = initialValue;
      node.id = id;
      node.className = 'cockpit-text';
      this.cockpitNode.appendChild(node);
      this.textNodes[id] = {
        node: node,
        value: initialValue
      };
    },

    /**
     * Updates a text item
     *
     * @param {string} id The identifier of this text item
     * @param {string} value The value to be displayed
     */
    updateText: function(id, value){
      var textItem = this.textNodes[id];
      if(textItem.value !== value){
        textItem.node[this.textProperty] = value;
      }
    }
  };

  global.Cockpit = Cockpit;

})(this);
